cmd/junitreport/pkg/parser/stack/parser.go

```diff
@@ -0,0 +1,95 @@
+func NewParser(builder builder.TestSuitesBuilder, testParser parser.TestDataParser, suiteParser parser.TestSuiteDataParser) parser.TestOutputParser {
+	return &testOutputParser{
+		builder:     builder,
+		testParser:  testParser,
+		suiteParser: suiteParser,
+	}
+}
+
+// testOutputParser uses a stack to parse test output. Critical assumptions that this parser makes are:
+//   1 - packages may be nested but tests may not
+//   2 - no package declarations will occur within the boundaries of a test
+//   3 - all tests and packages are fully bounded by a start and result line
+//   4 - if a package or test declaration occurs after the start of a package but before it's result,
+//       the sub-package's or member test's result line will occur before that of the parent package
+//       i.e. any test or package overlap will necessarily mean that one package's lines are a superset
+//       of any lines of tests or other packages overlapping with it
+//   5 - any text in the input file that doesn't match the above regex is necessarily the output of the
+//       current test
+type testOutputParser struct {
+	builder builder.TestSuitesBuilder
+
+	testParser  parser.TestDataParser
+	suiteParser parser.TestSuiteDataParser
+}
+
+// Parse parses output syntax of a specific class, the assumptions of which are outlined in the struct definition.
+// The specific boundary markers and metadata encoding is free to vary as long as it can be expressed as
+func (p *testOutputParser) Parse(input *bufio.Scanner) (*api.TestSuites, error) {
+	inProgress := NewTestSuiteDataStack()
+
+	currentTestData := &parser.TestData{}
+
+	for input.Scan() {
+		line := input.Text()
+
+		switch {
+		case p.testParser.ContainsIdentifier(line):
+			currentTestData.Identifiers = append(currentTestData.Identifiers, line)
+			fallthrough
+		case p.testParser.ContainsResult(line):
+			currentTestData.Result = line
+			fallthrough
+		case p.testParser.ContainsDuration(line):
+			currentTestData.Duration = line
+			fallthrough
+		case p.testParser.ContainsMessage(line):
+			fallthrough
+		case p.suiteParser.ContainsIdentifier(line):
+			// if we encounter the beginning of a suite, we create a new suite to be considered and
+			// add it to the head of our in progress package stack
+			inProgress.Push(&parser.TestSuiteData{
+				Identifier: line,
+			})
+			fallthrough
+		case p.suiteParser.MarksCompletion(line):
+			// if we encounter the end of a suite, we finalize the suite at the head of the in progress
+			// suite stack and remove it from the stack
+			currentSuite, err := inProgress.Pop().BuildTestSuite(p.suiteParser)
+			if err != nil {
+				return nil, err
+			}
+
+			p.builder.AddSuite(currentSuite)
+			fallthrough
+		default:
+			// if we did not encounter the beginning or end of a package or test, we are finding test output
+			currentTestData.Output = append(currentTestData.Output, line)
+		}
+
+		if currentTestData.IsComplete() {
```

Comment:
This feels like it could miss cases where we got a super clear indication a new test was starting (`=== RUN`) but misparsed the results line (maybe they change the duration format to nanoseconds or something). Seems like we should give the test data or test parser a chance to indicate it is complete, and close out the current test case, AND give the test parser a chance to indicate a new test is starting.

Initializing test and suite data with "unknown" name, result, duration, etc, populating those things as we know them, and pushing results into the current suite and moving on with a new test data if we see a new test starting makes us more robust to parse or output errors.